File: robot_ws/src/basic_pkg/basic_pkg/helloworld_publisher.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile
from std_msgs.msg import String

class HelloworldPublisher(Node):
    def __init__(self):
        super().__init__('helloworld_publisher')
        qos_profile = QoSProfile(depth=10) # buffer max 10 data in case of communication failure
        self.helloworld_publisher = self.create_publisher(String, 'helloworld', qos_profile) # topic name is /helloworld
        self.timer = self.create_timer(1, self.publish_helloworld_msg) ## execute callback function every 1 second
        self.count = 0 
    
    def publish_helloworld_msg(self): # callback function
        msg = String()
        msg.data = f'Hello World: {self.count}'
        self.helloworld_publisher.publish(msg) # publish msg
        self.get_logger().info(f'Published message: {msg.data}') # similar functionality to print function
        self.count += 1
    
def main(args=None):
    rclpy.init(args=args)
    node = HelloworldPublisher()
    
    try:
        rclpy.spin(node) # execute callback function repeated times
    except KeyboardInterrupt:
        node.get_logger().info('Keyboard Interrupt (SIGINT)')
    finally:
        node.destroy_node()
        rclpy.shutdown() # finish node

if __name__ == '__main__':
    main()

        
    

File: robot_ws/src/basic_pkg_cpp/src/helloworld_publisher.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp" 
#include "std_msgs/msg/string.hpp" // string type interface

using namespace std::chrono_literals; // time namespace

class HelloworldPublisher : public rclcpp::Node // inherit Node class
{
public:
    HelloworldPublisher()
    : Node("helloworld_publisher"), count_(0) // Node class constructor, count init
    {
        auto qos_profile = rclcpp::QoS(rclcpp::KeepLast(10)); // QOS setting depth=10
        helloworld_publisher_ = this->create_publisher<std_msgs::msg::String>( // topic type
            "helloworld", qos_profile // topic name, QoS setting
        ); // create publisher
        timer_ = this->create_wall_timer( // callback function every 1 second
            1s, std::bind(&HelloworldPublisher::publish_helloworld_msg, this)
        );
    }
private:
    void publish_helloworld_msg() // callback function
    {
        auto msg = std_msgs::msg::String();
        msg.data = "Hello World: " + std::to_string(count_++);
        RCLCPP_INFO(this->get_logger(), "Message published: '%s", msg.data.c_str()); // logging
        helloworld_publisher_->publish(msg);
    }
    // private member variable initialization
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr helloworld_publisher_;
    size_t count_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<HelloworldPublisher>(); // shared pointer node
    rclcpp::spin(node); // run callback function inside node
    rclcpp::shutdown();
    return 0;
}